### PR TITLE
add DataCollatorForDenoisingTasks

### LINKED ISSUE #20: Slowness created by pile.py
The download manager should cache copies of the the pile. It should also use lm_dataformat.

See here:
https://github.com/leogao2/lm_dataformat

I do not think streaming is currently implemented correctly. As it stands, we keep trying to decompress the file every time we generate. 

lm_dataformat also supports multithreaded streaming.

### PR DESCRIPTION
**Pull request by mattbui**
## Added
- New data collator class `DataCollatorForDenoisingTasks`, which is more accurate to https://github.com/pytorch/fairseq/blob/1bba712622b8ae4efb3eb793a8a40da386fe11d0/fairseq/data/denoising_dataset.py
	- Add shift_tokens_right for decoder_input_ids
- Pytorch DataLoader for batching
- `SentenceTokenize` will split long texts into smaller chunks when encountering long input text to speed up the tokenizer.
- Add `data_files` parameters for load pile dataset locally:
```python
from datasets import load_dataset
data_files={
	"train": "./data/train_files",
	"validation": "./data/...zst",
	"test": "./data/...zst"
}
dataset = load_dataset("./pile.py", data_files=data_files) # set streaming=True here also fine, it will load the dataset on the fly.
```

## Changes
- Replace grouping texts and manual iterate for batching with PyTorch DataLoader
- Reduce shuffle_buffer_size for better data loading startup, `shuffer_buffer_size = 1000` is still make start-up time slow, but it's good enough

Here's the new data flow:
```python
dataset = load_dataset(...)

sent_tok = SentenceTokenize()
dataset = dataset.map(sent_tok, batched=True, batch_size=1) # batched=True as the number of examples changed due to split up long doc

def tokenizer_fn(...):
	...
dataset = dataset.map(tokenizer_fn, bathed=True, batch_size=1)

data_collator = DataCollatorForDenoisingTasks(tokenizer)

iter = BackgroundGenerator(DataLoader(dataset, batch_size=2, collate_fn=data_collator))
```

Fixed #20

### CODE REVIEW

**Comment by mattbui** [comment]
@LouisCastricato please also take a look

**Comment by mattbui** [comment]
@morganmcg1 @tylerjthomas9 @LouisCastricato This PR is done, I updated the descriptions about the changes.

**Comment by Shashi456** [comment]
Is the .zst datafile requirement mandatory? or would it work with the decompressed jsonline files as well?

**Comment by LouisCastricato** [comment]
Do not get rid of zst. It's a death sentence without it

**Review by LouisCastricato — APPROVED** [review]

File: run_dnlm_flax.py [hunk 1]
```diff
@@ -397,57 +341,50 @@ def write_eval_metric(summary_writer, eval_metrics, step):
...
-      pad_token="<pad>",
-      cls_token="",
-      mask_token="<mask>",
-      eos_token="<s>",
-      bos_token="</s>"
+        model_args.tokenizer_name,
+        unk_token="<unk>",
+        sep_token="</s>",
+        pad_token="<pad>",
+        cls_token="<s>",
```
Comment (morganmcg1) [inline run_dnlm_flax.py:348]: Do we need `cls_token` ?

File: run_dnlm_flax.py [hunk 2]
```diff
@@ -397,57 +341,50 @@ def write_eval_metric(summary_writer, eval_metrics, step):
...
-    masking_collator = DataCollatorForTextInfilling(tokenizer)
-    tokenized_train_dataset = tokenized_train_dataset.map(masking_collator)
-    tokenized_eval_dataset = tokenized_eval_dataset.map(masking_collator)
+    tokenized_eval_dataset = sent_tokenized_eval_dataset.map(tokenize_function, batched=True, batch_size=1)
 
-    print("Collate deployed successfully.")
+    # Shuffle the training dataset
+    shuffled_train_dataset = train_dataset.shuffle(buffer_size=data_args.shuffle_buffer_size, seed=training_args.seed)
+
+    # items returned by shuffled dataset is nested in a list, we need to flatten it
```
Comment (morganmcg1) [inline run_dnlm_flax.py:372]: good catch

**Review by morganmcg1 — APPROVED** [review]
Thanks for all the work, looking to be in much better shape now

File: run_dnlm_flax.py (hunk 1 above)
Comment (mattbui) [reply run_dnlm_flax.py:348]: @morganmcg1 It's from the default config of `BartTokenizer`, so the tokenizer will add `<s>` and `</s>` tokens at the beginning and the end of the doc correctly. We can add it ourselves in the `SentenceTokenize` map() but it might be truncated by the tokenizer

Comment (mattbui) [reply run_dnlm_flax.py:348]: the cls_token here only act as a bos token for the encoder

Comment (morganmcg1) [reply run_dnlm_flax.py:348]: ah ok, makes sense 👍

**Comment by mattbui** [comment]
@LouisCastricato @morganmcg1 @tylerjthomas9 I corrected the evaluation data loading in this commit, [`2dcbc28` (#23)](https://github.com/morganmcg1/rotobart/pull/23/commits/2dcbc28317c183a77df330368be307431de4cbb1). We'll use the eval_iter directly as they're already batched and applied data_collator

**Review by tylerjthomas9 — APPROVED** [review]
